### DISCUSSION TITLE
Where to find paramteers defined in autotuner.json?

### DISCUSSION THREAD

**Discussion by dalyles** [The-OpenROAD-Project/OpenROAD-flow-scripts#1179] [Q&A] [answered]
Hello,

I am trying to do an example run of <code>autotuner.json</code> for the <code>gcd</code> circuit <code>nangate45</code>. I copied/pasted the <code>autotuner.json</code> file from the <code>sky130hd</code>  <code>gcd</code>. It seems to have a number of parameters listed which I will list here:

<code>_SDC_CLK_PERIOD</code>
<code>CORE_UTILIZATION</code>
<code>CORE_ASPECT_RATIO</code>
<code>CORE_MARGIN</code>
<code>CELL_PAD_IN_SITES_GLOBAL_PLACEMENT</code>
<code>CELL_PAD_IN_SITES_DETAIL_PLACEMENT</code>
<code>_FR_LAYER_ADJUST</code>
<code>PLACE_DENSITY_LB_ADDON</code>
<code>_PINS_DISTANCE</code>
<code>CTS_CLUSTER_SIZE</code>
<code>CTS_CLUSTER_DIAMETER</code>
<code>_FR_GR_OVERFLOW</code>

So my question is, where do you find these parameters? I assume you have to set them to the values determined by the Autotuner. And also why do some parameters start with an underscore?

Thanks

**Comment by vijayank88** [comment]
@vvbandeira Can you please respond to this?

**Comment by vvbandeira** [comment, marked as answer]
@dalyles we do not have a docs to point to you at this moment. But in short:
- Variables without underscore are any variable you can give the `make` command. For these, we have a non-exhaustive list [here](https://openroad-flow-scripts.readthedocs.io/en/latest/user/FlowVariables.html).
- Variables with an underscore are special cases, mainly values that need to be written in the `sdc` and `fastroute.tcl` files. You can look at the code that handles them to better understand their behaviour [here](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/flow/util/distributed.py#L362) and [here](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/flow/util/distributed.py#L401).

**Reply by dalyles** [reply]
Thanks,

I also have a followup question. 
I successfully ran the autotuner for <code>gcd</code>  in <code>nangate45</code>.
However, the run ID with the optimal parameters, <code>84eb1d92</code>,has an incomplete list of logs that you would find in <code>/OpenROAD-flow-scripts/flow/logs/nangate45/gcd/test-tune-2023-06-26-17-27-48/variant-AutoTunerBase-84eb1d92-or</code> for example.
<img width="711" alt="image" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/95712444/d7b26ea5-e965-4e65-a783-3f3afdf05775">

Right here shows the <code>.json</code> file with the corresponding run ID
<img width="1056" alt="image" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/95712444/b3470c12-6553-4864-b541-3bd2e524a998">

Is this behavior expected or is the tool falsely giving optimal parameters without regards to where a run is completed?

Here is a screenshot of <code>2_2_floorplan_io.log</code> for reference
<img width="719" alt="image" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/95712444/5ae86f0f-c559-4a2b-8a7b-1d88f2bb5139">

Thanks

**Reply by vvbandeira** [reply]
This is a bug, please file an issue with the required steps and files to reproduce.

**Reply by vijayank88** [reply]
@dalyles 
can you share the command and `autotuner.json`  file used?

**Reply by dalyles** [reply]
Sure, the command I used is:

<code>python3 distributed.py --design gcd --platform nangate45  --config ../designs/nangate45/gcd/autotuner.json tune</code>

And here's the <code>autotuner.json</code> file: (for some reason github doesn't allow me to attach json files)

<pre><code>{
    "_SDC_FILE_PATH": "constraint.sdc",
    "_SDC_CLK_PERIOD": {
        "type": "float",
        "minmax": [
            1.0,
            4.3647
        ],
        "step": 0
    },
    "CORE_UTILIZATION": {
        "type": "int",
        "minmax": [
            5,
            100
        ],
        "step": 1
    },
    "CORE_ASPECT_RATIO": {
        "type": "float",
        "minmax": [
            0.5,
            2.0
        ],
        "step": 0
    },
    "CORE_MARGIN": {
        "type": "int",
        "minmax": [
            2,
            2
        ],
        "step": 0
    },
    "CELL_PAD_IN_SITES_GLOBAL_PLACEMENT": {
        "type": "int",
        "minmax": [
            0,
            5
        ],
        "step": 1
    },
    "CELL_PAD_IN_SITES_DETAIL_PLACEMENT": {
        "type": "int",
        "minmax": [
            0,
            5
        ],
        "step": 1
    },
    "_FR_LAYER_ADJUST": {
        "type": "float",
        "minmax": [
            0.1,
            0.7
        ],
        "step": 0
    },
    "PLACE_DENSITY_LB_ADDON": {
        "type": "float",
        "minmax": [
            0.0,
            0.99
        ],
        "step": 0
    },
    "_PINS_DISTANCE": {
        "type": "int",
        "minmax": [
            1,
            4
        ],
        "step": 1
    },
    "CTS_CLUSTER_SIZE": {
        "type": "int",
        "minmax": [
            10,
            200
        ],
        "step": 1
    },
    "CTS_CLUSTER_DIAMETER": {
        "type": "int",
        "minmax": [
            20,
            400
        ],
        "step": 1
    },
    "_FR_FILE_PATH": "../../../platforms/nangate45/fastroute.tcl",
    "_FR_GR_OVERFLOW": {
        "type": "int",
        "minmax": [
            1,
            1
        ],
        "step": 0
    }
}</code></pre>

**Comment by vijayank88** [comment]
@dalyles 
See my screenshot used with your `autotuner.json`. I am getting all logs relevant to optimal json parameters.
![Screenshot from 2023-06-28 17-13-59](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/85787913/92c981d8-317b-4a37-be7b-5830b791f369)

**Reply by dalyles** [reply]
Hmm strange...so it looks like it successfully generates the logs on your machine when you use the same exact commands to reproduce the results.

**Reply by dalyles** [reply]
I did a rerun and it mysteriously has all the logfiles now.
